Mk transifex order independence

mkangia opened the pull request:
[Modules and forms order Independence]
According to current integration, the resources on Transifex(A resource is a single PO file on Transifex) are kept according to how its on HQ like 
module1, module1_form1 i.e dependent on the order of modules and forms.

In case the module/forms are rearranged on HQ, it would conflict with transifex. For example:
say, 
module1 and module2 were pushed to transifex as `module1` and `module2`
module1 and module2 are interchanged on HQ.
now module1 on HQ becomes `module2` on Transifex and vice versa and hence when updated translations would be pushed on HQ, it would wipe out uncommon translations on Transifex i.e the ones that were present in the module mapped earlier[module1] and are not present in new module[module2].

To avoid the scenario above,
this PR updates how we maintain resource slugs on Transifex. Instead of keeping module1 we can now keep module_moduleUniqueID or module1_form1 as form_formUniqueID so that even if the order is changed, the resource slugs would not change and hence it would map to the same module/form irrespective of its place in HQ App.

[Questions within a form and case properties within a module]
[context]
A context along with a translation is kept to uniquely identify it. This is useful in case there are two messages with the same source, say "Name" but used at different places. We also used "index" of the translation within the excel sheet structure as a part of the context along with other identifier. For example for a module sheet, the first translation has case_property `name` with list_or_detail as `list`, hence the context is `0:name:list`. It was more of a safe keeping.

[change done]
Having index in the context for a translation invalidates the translation if its index changes because it then changes the context, even though its the same message to be translated.
It was added during development/testing but the possibility of it being useful is hopefully skim in real world. So considering it wipes out quite a lot on Transifex even when a single question is removed/added in between a form, the cost is high and hence its better to not use it now.
In case there is a duplicate, a message like this is shown to the user for the corresponding resource
```
A duplicate msgid was detected (Name). Use a unique msgid or a unique msgctxt to differentiate (the existing msgctxt `name:list` is a duplicate one)
```
the only work around as of now is to remove the duplicate in the app itself.
We do a keep a different label id for questions in the form so there should not be any issue. The only place i can think of is when one uses the same case property in a module's "Case List" or "Case Detail", would cause this error.

mkangia committed:
ensure that order is maintained when creating PO files

mkangia committed:
generate module/form sheets with unique id

- generate module/form resource slugs with unique id
- maintain module/form name for adding human redable name to resources

mkangia committed:
update parser

Add ability to parse resources with module/form unique ID in their slugs

mkangia committed:
lint

mkangia committed:
remove irrelevant variable for unique id

mkangia committed:
update form resource name

use module name along with form name for resources on transifex so that two forms with same name but in different modules can be differentiated

mkangia committed:
fix fetching resource name

use unique id to map reosurce name instead to avoid issues when using/not using version postfix

millerdev commented on corehq/apps/app_manager/models.py:
```diff
@@ -2523,6 +2522,14 @@ def get_form(self, i):
+            if form.unique_id == unique_id:
+                return index
+        error = _("Could not find form with ID='{unique_id}' in module '{app_name}'.").format(
+            app_name=self.name, unique_id=unique_id)
```
Is `app_name` the name of the app or the module? The error text makes it sound like a module name, but the variable name makes it sound like an app name. It looks like either the variable or the error text could be changed to clear that up. Guessing from the code below in `get_module_index` that the error text should be changed.

mkangia replied:
Yup, should be `module_name`

millerdev commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -70,19 +72,38 @@ def client(self):
+
+        :param resource_slug: like module_moduleUniqueID
+        """
+        if MODULES_AND_FORMS_SHEET_NAME in resource_slug:
```
Could this be changed to `if resource_slug == MODULES_AND_FORMS_SHEET_NAME:`? It seems like that would be more consistent other code in this PR.

mkangia replied:
Its possible that resource_slug contains "_v12" in the end in case the translations files are kept based on version and so the check is a "in". The other way is to split first and then check. Can do that if that is better to understand.

millerdev replied:
I see. So you could say `resource_slug.startswith(MODULES_AND_FORMS_SHEET_NAME + "_v")` and that would be less likely to cause a false positive, right? It's probably fine as is unless you think there is any chance that `resource_slug` could somehow contain `MODULES_AND_FORMS_SHEET_NAME` but not be the sheet you want to match.

mkangia replied:
> somehow contain MODULES_AND_FORMS_SHEET_NAME but not be the sheet

Not possible. We control it. It can be either
module_moduleUniqueID(_v12) 
form_formUniqueID(_v12)
Module_and_forms(_v12)

mkangia replied:
.. well.. even that step is controlled by HQ. 
This is not a user input.
The spreadsheet like structure is generated by HQ, [here](https://github.com/dimagi/commcare-hq/blob/42e5e59fb9f21d5161fe2460767eda81b2b820b5/corehq/apps/app_manager/app_translations/app_translations.py#L299-L298).

millerdev replied:
So it's not an actual spreadsheet, just a dict that has a similar interface. I didn't realize that. Ignore what I said.

millerdev commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -70,19 +72,38 @@ def client(self):
+        module_or_form_unique_id = resource_slug.split('_')[1]
+        resource_name_in_all_langs = self.transifex_po_file_generator.slug_to_name[module_or_form_unique_id]
+        return resource_name_in_all_langs.get(self.transifex_project_source_lang,
+                                              resource_name_in_all_langs.get('en', resource_slug))
```
Curious why `'en'` is used here? Is there a default language for the app or domain? Would it make sense to use that instead?

mkangia replied:
Nothing in particular. Just used english as default, though its highly unlikely this would ever happen since the transifex project source lang should always be same as default language on HQ app for transifex integration to work, which in the app, should always be present. 
So just used "en" in case someone tries something out of the box.

mkangia commented:
@millerdev 
also Daniel, thanks for the feedback. Was still working on the PR so didn't add reviewers yet, sorry you had to look at some unfinished work, though its almost ready.

mkangia committed:
fix name

mkangia committed:
Dont use index in the context of translation

mkangia commented:
@millerdev this is ready now. just had a small though critical change. Update the description in the PR. The second block is the new one i added.

mkangia commented on corehq/apps/app_manager/app_translations/generators.py:
```diff
@@ -92,7 +96,63 @@ def _set_sheet_name_to_module_or_form_mapping(self, all_module_and_form_details)
     def _get_filename(self, sheet_name):
+        """
+        receive sheet name in HQ format and return the name that should be used
+        to upload on transifex along with module/form unique ID and version postfix
```
> Any idea how hard it'd be to switch that to use unique IDs too

Use unique IDs in HQ?
I think It makes more sense for user to see "Module0_form1" as the sheet name instead of ids in the excel sheet.

> to make the intermediate format contain both IDs? 

You mean hq to give both ids already? i am not sure how useful ID would be anywhere else so i would avoid adding that to HQ.

mkangia committed:
Resolve conflicts

mkangia committed:
fix issues due to master merge 1

Accomodate for changes done https://github.com/dimagi/commcare-hq/pull/21723/
transifex_po_file_generator was removed from object and renamed

mkangia committed:
fix issues due to merge master 2

parser was broken bcoz transifex_po_file_generator does not exist anymore.
Moved app_id_to_build to transifex and provide that to AppTranslationGenerator and then to parser

mkangia committed:
fix issues with return and module sheet name generated

mkangia commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -52,11 +52,12 @@ def send_translation_files(self):
             self.exclude_if_default, self.use_version_postfix)
-        with PoFileGenerator(app_trans_generator.translations, app_trans_generator.metadata) as po_file_generator:
+        with PoFileGenerator(self.app_trans_generator.translations,
+                             self.app_trans_generator.metadata) as po_file_generator:
```
@millerdev @esoergel 
Considering the great conversations we have had last week,
Curious if you think its fine to 
just put `app_trans_generator` on `self` so that its available later or 
pass `app_trans_generator` in the argument and then use `app_trans_generator. slug_to_name` or
pass `app_trans_generator.slug_to_name` as the argument itself for use later?

millerdev replied:
If you assign it to `self` then it becomes an order-dependent resource and will make maintaining this class more difficult because the life-cycle of that attribute is not immediately clear. My recommendation would be to pass it to the method(s) that need it rather than assign it to `self`.

esoergel replied:
As you suggested, if you do need to keep around part of a class, I'd store just the simple attributes you need, such as in this case `slug_to_name`.  That still runs into the issue @millerdev mentioned - pylint gives a warning for class attributes defined outside of `__init__` for that reason, actually.

mkangia commented:
@millerdev @esoergel 
had to adjust to changes done in https://github.com/dimagi/commcare-hq/pull/21723. Request you to take one more look.

mkangia committed:
pass object as argument instead
